src/type-coercions.md

```diff
@@ -190,40 +185,92 @@ r[coerce.unsize]
+
+r[coerce.unsize.traits]
+Three traits, [`Unsize`], [`CoerceUnsized`], and [`PinCoerceUnsized`] are used to assist in this process and expose it for library use.
+
+r[coerce.unsize.traits.unsize]
+[`Unsize`] represents the fact that the target type is layout compatible with the source type and the pointer metadata of the target type can be derived from the metadata of the source, meaning that a pointer to the source type can be converted to a pointer to the target type. For example `[T; N]` implements `Unsize<[T]>` meaning that you can *unsize* former into the later, allowing coercions such as `&[T; N] -> &[T]`.
+
+r[coerce.unsize.traits.coerce-unsized]
+[`CoerceUnsized`] represents the fact that a pointer-like type can be coerced to another pointer-like type, due to `Unsize` being implemented for their pointees. For example, `&T` implements `CoerceUnsized<&U>` when `T: Unsize<U>`.
+
+r[coerce.unsize.traits.pin-coerce-unsized]
+[`PinCoerceUnsized`] is an unsafe marker trait for pointer-like types unsized coercion of which does not break [`Pin`] guarantees. It is a requirement of the `CoerceUnsized` implementation for `Pin`. That is, `&D: PinCoerceUnsized` implies `Pin<&T>: CoerceUnsized<Pin<&U>>`.
+
+The following implementations of [`Unsize`] are built-in:
 
 r[coerce.unsize.slice]
-* `[T; n]` to `[T]`.
+* `[T; n]: Unsize<[T]>`.
 
 r[coerce.unsize.trait-object]
-* `T` to `dyn U`, when `T` implements `U + Sized`, and `U` is [dyn compatible].
+* `T: Unsize<dyn U>`, when `T` implements `U + Sized`, and `U` is [dyn compatible].
 
 r[coerce.unsize.trait-upcast]
-* `dyn T` to `dyn U`, when `U` is one of `T`'s [supertraits].
+* `dyn T: Unsize<dyn U>`, when `U` is one of `T`'s [supertraits].
     * This allows dropping auto traits, i.e. `dyn T + Auto` to `dyn U` is allowed.
     * This allows adding auto traits if the principal trait has the auto trait as a super trait, i.e. given `trait T: U + Send {}`, `dyn T` to `dyn T + Send` or to `dyn U + Send` coercions are allowed.
 
-r[coerce.unsized.composite]
-* `Foo<..., T, ...>` to `Foo<..., U, ...>`, when:
-    * `Foo` is a struct.
+r[coerce.unsize.composite]
+* `S<..., T, ...>: Unsize<S<..., U, ...>>`, when:
+    * `S` is a struct.
     * `T` implements `Unsize<U>`.
-    * The last field of `Foo` has a type involving `T`.
-    * If that field has type `Bar<T>`, then `Bar<T>` implements `Unsize<Bar<U>>`.
-    * T is not part of the type of any other fields.
+    * The last field of `S` has a type involving `T`. i.e. it's either of `T` or `C<..., T, ...>` where `C` is a type constructor and `T` is only present in it once (`C<T, T>` is disallowed).
+    * The last field is the *only* field which type involves `T`.
+    * The type of the last field implements `Unsize<F>` where `F` is the same type with `T` replaced by `U`. i.e. if the field has type `Bar<T>`, then `Bar<T>` implements `Unsize<Bar<U>>`.
+
+r[coerce.unsize.pointer]
+A type `S<..., T, ...>` can implement `CoerceUnsized<S<..., U, ...>>` if
```

Comment:
see https://github.com/rust-lang/rust/blob/2cd37831b0706c9f2d6e7e20eb556dc7548bf732/compiler/rustc_next_trait_solver/src/solve/trait_goals.rs#L1048

unsizing can change an arbitrary number of struct type/const args as long all of them are only mentioned in the last field and that last field impls `CoerceUnsized`

> The last field of `S` has a type involving `T`. i.e. it's either of `T` or `C<..., T, ...>` where `C` is a type constructor and `T` is only present in it once (`C<T, T>` is disallowed).

Where do you get the "`T` is only present once" requirement from?

Reply:
wait, confused by the difference between `coerce.unsize.composite` and `coerce.unsize.pointer` :thinking: using

Reply:
an example which disagrees with the current docs
```rust
trait Trait<T, U> {
    type Assoc: ?Sized;
}

impl<T> Trait<T, u32> for T {
    type Assoc = [u8; 4];
}

impl<T> Trait<T, u16> for T {
    type Assoc = [u8];
}

struct Foo<T: Trait<T, U>, U>(<T as Trait<T, U>>::Assoc);

fn main() {
    let x: &Foo<(), u32> = &Foo([0; 4]);
    let y: &Foo<String, u16> = x;
}
```

Reply:
I fixed both of these to be in-line with the compiler. Could you check that I didn't mess it up again?

Reply:
> Where do you get the "T is only present once" requirement from?

I think this was an old limitation which isn't true anymore. (or I interpreted something incorrectly in the past?)